### PR TITLE
Update no trainer scripts for language modeling and image classification examples

### LINKED ISSUE #18437: Update no_trainer scripts to include gather_for_metrics
### Feature request

🤗 Accelerate has a wrapper to help with distributed metric calculation (a tough problem!), and the `no_trainer` scripts should be updated to include it! 

An example can be seen [here](https://github.com/huggingface/accelerate/blob/main/examples/nlp_example.py#L163-L169), below is an example diff of what the integration would look like:

```diff
-             predictions, references = accelerator.gather((predictions, batch["labels"]))
-             # If we are in a multiprocess environment, the last batch has duplicates
-             if accelerator.num_processes > 1:
-                 if step == len(eval_dataloader) - 1:
-                     predictions = predictions[: len(eval_dataloader.dataset) - samples_seen]
-                     references = references[: len(eval_dataloader.dataset) - samples_seen]
-                 else:
-                     samples_seen += references.shape[0]
+             predictions, references = accelerator.gather_for_metrics((predictions, batch["labels"])) 
```

The list of available scripts to update include:

- [x] examples/pytorch/image-classification/run_image_classification_no_trainer.py
- [x] examples/pytorch/language-modeling/run_clm_no_trainer.py
- [x] examples/pytorch/language-modeling/run_mlm_no_trainer.py
- [x] examples/pytorch/multiple-choice/run_swag_no_trainer.py
- [x] examples/pytorch/question-answering/run_qa_beam_search_no_trainer.py
- [x] examples/pytorch/question_answering/run_qa_no_trainer.py
- [x] examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py
- [x] examples/pytorch/speech-pretraining/run_wav2vec2_pretraining_no_trainer.py
- [ ] examples/pytorch/summarization/run_summarization_no_trainer.py

### Motivation

This is a great first issue for someone who wants to learn how to use some of the latest bits in Accelerate and get an easy beginner contribution to the library 🤗 

### Your contribution

If you decide to pick up this issue, feel free to ping myself (@muellerzr), @sgugger, or @pacman100 to review 🤗

### PR DESCRIPTION
**Pull request by nandwalritik**
# What does this PR do?

Fixes #18437 
Updated no_trainer scripts for `examples/pytorch/image-classification/run_image_classification_no_trainer.py`, `examples/pytorch/language-modeling/run_clm_no_trainer.py` and  `examples/pytorch/language-modeling/run_mlm_no_trainer.py` to include`gather_for_metrics`.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@muellerzr @sgugger 
Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: examples/pytorch/language-modeling/run_clm_no_trainer.py [hunk 1]
```diff
@@ -597,7 +597,7 @@ def group_texts(examples):
                 outputs = model(**batch)
 
             loss = outputs.loss
-            losses.append(accelerator.gather(loss.repeat(args.per_device_eval_batch_size)))
+            losses.append(accelerator.gather_for_metrics(loss.repeat(args.per_device_eval_batch_size)))
```
Comment (sgugger) [inline examples/pytorch/language-modeling/run_clm_no_trainer.py:600]: You then need to remove the truncation at line 603 below :-)

File: examples/pytorch/language-modeling/run_mlm_no_trainer.py [hunk 2]
```diff
@@ -642,7 +642,7 @@ def group_texts(examples):
                 outputs = model(**batch)
 
             loss = outputs.loss
-            losses.append(accelerator.gather(loss.repeat(args.per_device_eval_batch_size)))
+            losses.append(accelerator.gather_for_metrics(loss.repeat(args.per_device_eval_batch_size)))
```
Comment (sgugger) [inline examples/pytorch/language-modeling/run_mlm_no_trainer.py:645]: And at line 648 here

**Review by sgugger — COMMENTED** [review]
Thanks a lot for working on this! Left two comments on the LM examples.

File: examples/pytorch/language-modeling/run_mlm_no_trainer.py (hunk 2 above)
Comment (nandwalritik) [reply examples/pytorch/language-modeling/run_mlm_no_trainer.py:645]: fixed.

File: examples/pytorch/language-modeling/run_clm_no_trainer.py (hunk 1 above)
Comment (nandwalritik) [reply examples/pytorch/language-modeling/run_clm_no_trainer.py:600]: fixed.

**Review by muellerzr — APPROVED** [review]
Great work! 😄 👍

**Review by sgugger — APPROVED** [review]

**Comment by sgugger** [comment]
Thanks again for your contribution!
